### ISSUE TITLE
query::unread "loses" articles on reload

### ISSUE THREAD

**Issue by allanwind** [newsboat/newsboat#2150]
**Newsboat version (copy the output of `newsboat -v` or the first line of `git show`)**: r2.27-80-ga42f.

I usually reproduce against master, but newsboat no longer build on debian stable:

error[E0658]: arbitrary expressions in key-value attributes are unstable                                               
   --> /home/allan/.cargo/registry/src/github.com-1ecc6299db9ec823/libz-sys-1.1.8/src/lib.rs:115:19                    
    |                                                                                                                  
115 |     #[link_name = zng_prefix!(adler32)]                                                                          
    |                   ^^^^^^^^^^^^^^^^^^^^                                                                           
    |                                                                                                                  
    = note: see issue #78835 <https://github.com/rust-lang/rust/issues/78835> for more information

Config file (copy from ~/.newsboat/config or ~/.config/newsboat/config):

```
article-sort-order title
articlelist-title-format "%U"
bind-key J next-feed
bind-key K prev-feed
bind-key j next
bind-key k prev
bind-key o open-in-browser-and-mark-read articlelist
confirm-mark-feed-read no
datetime-format "%F %T"
error-log ~/.local/share/newsboat/error.log
feedlist-format "%4i %n %11u %L - %t"
ignore-mode display
keep-articles-days 365
prepopulate-query-feeds yes
reload-threads 32
run-on-startup reload-all;open
show-read-articles no
show-read-feeds no
unbind-key C
```

The main one of interest is `show-read-articles no` combined with this url `"query:Unread:unread = \"yes\""`

Steps to reproduce the issue:

1. newsboat and wait for reload-all to finish
2. quit
3. newsboat
4. read the first article
5. reload-all (R)
6. read articles is no longer in query feed, and you cannot find it with toggle-article-read.

**Other info you think is relevant**:

This is a pretty esoteric bug but if you configure auto-reload would be surprising.  I don't know if this is a duplicate of #384 (closed), #387 or #388.  It would be relevant for #1730 if we get there.

**Comment by Minoru** [comment]
> newsboat no longer build on debian stable:

Support for Rust 1.48.0 ended with Newsboat 2.25. I should probably document the policy, but for now you can see https://github.com/newsboat/newsboat/issues/709. The gist of it is that the crates we depend on don't usually support older Rust versions, so neither can we.

As for the situation you describe, I'm not entirely sure that it's a bug. The query feed you created explicitly asks for unread articles, so reloading that feed populates it with unread articles -- seems like the expected behaviour? OTOH I do understand that articles disappearing right in front of the user's eyes in not a great UX.

Perhaps we should take cue from how we handle deletions: `delete-article` marks article as deleted, but keeps it in the list; then `purge-deleted` removes all deleted articles from the list. The list is also purged when the dialog is closed and opened again, or when the feed is reloaded. An important difference UX-wise is that deleted articles are displayed with a "D" in one of the columns, so it's easy to understand that they're deleted; how should we mark the entries in a query feed that no longer match the query's filter expression? In your example it's simple (unread articles have "N", others don't), but we need a more generic solution because query feeds can be different.

**Comment by allanwind** [comment]
> ... you can see https://github.com/newsboat/newsboat/issues/709.

@AlexanderBatischev Nice write-up.

>As for the situation you describe, I'm not entirely sure that it's a bug. The query feed you created explicitly asks for unread articles, so reloading that feed populates it with unread articles -- seems like the expected behaviour? OTOH I do understand that articles disappearing right in front of the user's eyes in not a great UX.

Exactly.  Feel free to reject it as a non-issue or classify
it as an enhancement.

>Perhaps we should take cue from how we handle deletions: `delete-article` marks article as deleted, but keeps it in the list; then `purge-deleted` removes all deleted articles from the list. The list is also purged when the dialog is closed and opened again, or when the feed is reloaded. An important difference UX-wise is that deleted articles are displayed with a "D" in one of the columns, so it's easy to understand that they're deleted; how should we mark the entries in a query feed that no longer match the query's filter expression? In your example it's simple (unread articles have "N", others don't), but we need a more generic solution because query feeds can be different.

In the case I described we already had the " "(?) instead of the 
"N" state and it was still surprising so I don't think it's a 
display issue.

You might be onto something about _explicitly_ modeling the
state of an item that was read but still being shown
(because the itemlist hasn't been reloaded).

Maybe it just comes down to needing the feature `fetch` that 
reloads a query/feed but leaves deleted and read articles on the 
list (`reload` = `fetch` + `purge-deleted`)?  `auto-reload` should 
probably be `fetch`.  It's up to the user if they want a `reload` or 
`fetch` the current query/feed.

**Comment by allanwind** [comment]
Related, if you mark an article read, then a reload happens the article will revert to unread.

**Comment by Minoru** [comment]
@allanwind I can't reproduce that. For me, already-read items disappear from the query feed, and they remain read in their respective feeds. Here's the urls file:

```
https://lore.kernel.org/lkml/CANiq72=Q+024x_Bb__RRT9e30QmcTKhzBB2=CmfukJTCjXVY-A@mail.gmail.com/t.atom
https://dpc.pw/feed
https://www.webtoons.com/en/challenge/whats-up-beanie/rss?title_no=295880
https://ihatereality.space/rss.xml
https://permlug.org/rss
https://fediverse.blog/@/bugaevc/atom.xml
https://www.fossjobs.net/rss/programmers/
https://dragan-stepanovic.github.io/feed.xml
https://ecosphere.press/feed/
https://blog.inex.rocks/feed.xml
https://forgefriends.org/index.xml
https://lore.kernel.org/lkml/20211206140313.5653-1-ojeda@kernel.org/t.atom
https://rsshub.app/telegram/channel/rks_tech_talk
https://lore.kernel.org/all/YdIfz+LMewetSaEB@gmail.com/t.atom
http://rmathew.com/atom.xml
https://lore.kernel.org/lkml/20211223141113.1240679-1-Jason@zx2c4.com/t.atom
https://lore.kernel.org/all/Ydm7ReZWQPrbIugn@gmail.com/t.atom
https://blog.nelhage.com/atom.xml
https://etfbaza.ru/feed/
https://www.eyrie.org/~eagle/journal/reviews.rss
https://warnewsupdates.blogspot.com/feeds/posts/default
https://heyokyay.com/feed/
https://draculadaily.substack.com/feed
https://nitter.fdn.fr/icfpcontest2022/rss
https://crt.sh/atom?q=fediverse.party
"query:Unread:unread = \"yes\""
```


And the config:

```
auto-reload yes
reload-time 2
```

How I tried to reproduce it (on a42ff2986948be42c68071e450083315feddbaf9):

1. run `rm -f /dev/shm/2150-new-cache.db` to ensure a clean slate
2. run `./newsboat --config-file=data/2150-new-config --url-file=data/2150-new-urls --cache-file=/dev/shm/2150-new-cache.db`
3. wait for the reload to finish
4. enter the query feed
5. press <kbd>N</kbd> 50 times to mark the first 50 articles read
6. wait a couple minutes for the reload to start
7. once the reload is finished, press <kbd>q</kbd> to go back to feedlist
8. note that some feeds now have counts like "(19/20)" or "(7/20)", indicating that some articles in them are marked read.

Am I reproducing it wrong?

Also, I'd appreciate separate issues for separate bugs, even if they're related. It helps keep each discussion focused. They can be easily linked with references in comments.

**Comment by allanwind** [comment]
Do you want me to file another ticket for the read reverts to unread?  I will supply steps to reproduce it where every you prefer.
